Applications/Spire/Source/Ui/SortedTableModel.cpp

```diff
@@ -0,0 +1,216 @@
+            if(m_comparator(m_source->at(rhs, i->m_index),
+              m_source->at(lhs, i->m_index))) {
+              return true;
+            } else if(m_comparator(m_source->at(lhs, i->m_index),
+              m_source->at(rhs, i->m_index))) {
+              return false;
+            }
+          } else {
+            return false;
+          }
+        }
+        return false;
+      });
+    for(auto i = m_mapping.begin(); i != m_mapping.end(); ++i) {
+      m_reverse_mapping[*i] = static_cast<int>(i - m_mapping.begin());
+    }
+  }
+}
+
+void SortedTableModel::translate(int direction, int row) {
+  auto row_index = m_reverse_mapping[row];
+  for(auto i = 0; i != static_cast<int>(m_mapping.size()); ++i) {
+    if(m_mapping[i] >= row) {
+      m_mapping[i] += direction;
+    }
+    if(m_reverse_mapping[i] >= row_index) {
+      m_reverse_mapping[i] += direction;
+    }
+  }
+}
+
+void SortedTableModel::on_operation(const Operation& operation) {
+  m_transaction.transact([&] {
+    visit(operation,
+      [&] (const AddOperation& operation) {
+        auto added_index = [&] {
+          if(operation.m_index >= static_cast<int>(m_mapping.size())) {
+            m_mapping.push_back(operation.m_index);
+            m_reverse_mapping.push_back(operation.m_index);
+            return operation.m_index;
+          } else {
+            auto row_index = m_reverse_mapping[operation.m_index];
+            translate(1, operation.m_index);
+            m_mapping.insert(m_mapping.begin() + row_index,
+              operation.m_index);
+            m_reverse_mapping.insert(
+              m_reverse_mapping.begin() + operation.m_index, row_index);
+            return row_index;
+          }
+        }();
+        if(m_order.empty() || m_order.front().m_order == Ordering::NONE) {
+          m_transaction.push(AddOperation{added_index});
+        } else {
+          sort();
+          m_transaction.push(AddOperation{m_reverse_mapping[operation.m_index]});
+        }
+      },
+      [&] (const RemoveOperation& operation) {
+        auto row_index = m_reverse_mapping[operation.m_index];
+        translate(-1, operation.m_index);
+        m_mapping.erase(m_mapping.begin() + row_index);
+        m_reverse_mapping.erase(m_reverse_mapping.begin() + operation.m_index);
+        m_transaction.push(RemoveOperation{row_index});
+      },
+      [&] (const UpdateOperation& operation) {
+        auto row_index = m_reverse_mapping[operation.m_row];
+        m_transaction.push(UpdateOperation{row_index, operation.m_column});
+        auto is_sorted_column = [&] {
+          for(auto i = m_order.begin(); i != m_order.end(); ++i) {
+            if(i->m_order == Ordering::NONE) {
+              break;
+            } else if(i->m_index == operation.m_column) {
+              return true;
+            }
+          }
+          return false;
+        }();
+        if(!is_sorted_column) {
+          return;
+        }
+        sort();
```

Comment:
Resorting on every update is way to inefficient, the complexity of a sort operation is ```Ω(n * log(n))``` which means every time a value is modified a full blown quasi-linear operation is performed. Consider a financial table that supports sorting like the one on this website:

https://www.tradingview.com/screener/

Now imagine if anytime any one of those numbers was updated, a full blown ```Ω(n * log(n))``` had to be performed. That table would effectively be unusable.